src/librustdoc/passes/collect_intra_doc_links/early.rs

```diff
@@ -0,0 +1,77 @@
+use rustc_ast as ast;
+use rustc_hir::def::Namespace::TypeNS;
+use rustc_hir::def_id::{LocalDefId, CRATE_DEF_ID};
+use rustc_interface::interface;
+use rustc_span::Span;
+
+use std::cell::RefCell;
+use std::mem;
+use std::rc::Rc;
+
+// Letting the resolver escape at the end of the function leads to inconsistencies between the
+// crates the TyCtxt sees and the resolver sees (because the resolver could load more crates
+// after escaping). Hopefully `IntraLinkCrateLoader` gets all the crates we need ...
+crate struct IntraLinkCrateLoader {
+    current_mod: LocalDefId,
+    crate resolver: Rc<RefCell<interface::BoxedResolver>>,
+}
+
+impl IntraLinkCrateLoader {
+    crate fn new(resolver: Rc<RefCell<interface::BoxedResolver>>) -> Self {
+        Self { current_mod: CRATE_DEF_ID, resolver }
+    }
+}
+
+impl IntraLinkCrateLoader {
+    crate fn visit_crate(&mut self, krate: &ast::Crate) {
+        assert_eq!(self.current_mod, CRATE_DEF_ID);
+        self.visit_attributes(&krate.attrs, krate.span);
+    }
+
+    fn visit_attributes(&mut self, attrs: &[ast::Attribute], span: Span) {
+        use crate::html::markdown::markdown_links;
+        use crate::passes::collect_intra_doc_links::preprocess_link;
+
+        // FIXME: this probably needs to consider inlining
```

Comment:
Ugh, apparently github dropped my comment because I wrote it on a commit instead of the PR itself. Do you happen to have it in your email? If not I can try to rewrite it from memory, it was discussing exactly how this bug could happen and how I don't know how to test it with a UI test.

Reply:
No, I don't see any traces of the extended version of this comment in my mail.

Reply:
Ok, it went something like this: "inlining" is refering to `doc(inline)`, not `#[inline]`. The test case for it would look something like:

```rust
// inner_crate
pub fn foo() {}

// middle_crate
/// Link to [inner_crate::foo]
pub fn bar() {}

// outer_crate
pub use middle_crate::bar;
```
In particular, the first time `inner_crate` will be loaded is when resolving the links on `bar`, which I don't think this pass will catch because it doesn't look at the attributes of the original definition, only of the re-export.

I'm not sure how to test this with a UI test because it requires three nested crates, and AFAIK `aux-build` only supports two.

Reply:
Seems to work fine when testing locally though :shrug: so this is probably ok.